migration to kubernetes

madmatah opened the pull request:
## Purpose

We want to replace Openshift with Kubernetes.

This is a meta PR related to a protected branch (`migration-to-k8s`).
The goal is to open sub-PRs that will be merged into this one.

## Proposal

- [x] Setup development environment based on kubernetes
- [x] Update core tasks to remove Openshift specific code
- [x] Update core_apps
- [x] Update apps
  - [x] ashley
  - [x] edxapp
  - [x] edxec
  - [x] elasticsearch
  - [x] etherpad
  - [x] flower
  - [x] forum
  - [x] hello
  - [x] kibana
  - [x] mailcatcher
  - [x] marsha
  - [x] nextcloud
  - [x] redis
  - [x] redis-sentinel
  - [x] richie
- [x] Remove deprecated apps that won't be migrated to k8s
  - [x] learninglocker
  - [x] moodlenet
- [x] Documentation
  - [x] Update README
  - [x] Temporarily remove outdated docs (to be brought back after revamp in a future PR)

sampaccoud committed:
✨(apps) allow defining resources requests for all pods

We need to improve pod scheduling on our Kubernetes clusters. By
setting resources requests without any limit, our pods are in
"Burstable" mode: we can define the minimum resources it requires upon
scheduling, while still allowing it to burst.

madmatah committed:
✨(playbook) introduce static services

Add support  for static services, with  fixed name, in addition  to the services
deployed with  a deployment stamp in  their name. It allows  applications inside
the cluster to use the services exposed by other applications, with a consistent
fqdn.

BREAKING  CHANGE :  blue/green related  tasks are  now updating  static services
instead of routes

madmatah committed:
✨(forum) add static service for elastic search

Add a static service for the forum elasticsearch, to be used easily by
other applications.

jmaupetit committed:
📝(project) update documentation using the CLI

Now that we have a CLI to use arnold, our documentation needed a solid
revamp to reflect recent changes.

The documentation no longer focuses on `edxapp` but the tool itself,
with a clear distinction between tool users and developers.

madmatah committed:
🐛(cli) add missing export to K8S_DOMAIN in arnold CLI

When the `--k8s-domain` is not provided to arnold CLI, a default value
is computed with the local IP address. This default value is not
exported to the docker container running ansible tasks though the
K8S_DOMAIN environment variable. As a result the base domain used in
routes is invalid.

madmatah committed:
🏗️(cli) use k8s instead of OKD in arnold CLI

We want to replace OKD with Kubernetes. The authentication mechanism
is sligthly different, because the commands `oc login` and `oc whoami
-t` do not exist with `kubectl`.

Arnold CLI has been updated to discover authentication information in
k8s configuration and service account secrets, as explained here :
https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-api/

Note that it is now mandatory to authenticate through a service
account to use arnold. Authenticating with a regular user is not
possible anymore, because it can be done in multiple ways (e.g. with
OpenID, via a kubectl plugin that need user interaction) and this
cannot be executed inside the docker container.

For local development purpose, the `oc cluster` command has been
replaced with k3d, an utility to start a k3s cluster within docker. By
default, k3d installs traefik as the default ingress controller. We
chose to replace it by the ingress-nginx controller instead, because
it's a popular choice of ingress controller and it is officially
supported by the kubernetes project. And unlike traeffik, by default,
the nginx controller provides a default self-signed certificate if
cert-manager is not installed.

madmatah committed:
🔧(core) replace OKD specific entities with K8s equivalent in core tasks

We want to deploy arnold applications on a Kubernetes instead of
Openshift. To do that, we must replace Openshift specific entities
with K8S equivalent in ansible tasks :

- Route(s) have been replaces with Ingress(es)
  The route*.yml.j2 template files are now ignored and should be named
  ingress*.yml.j2

- DeploymentConfig(s) have been replaced with Deployment(s)
  The dc*.yml.j2 template files are now ignored and should be named
  deploy*.yml.j2

For TLS certificates generation, we replaced openshift-acme controller
with cert-manager.

Important : we now assume that cert-manager is already installed on
the target K8S cluster. Since it is strongly recommended to install
only one cert-manager instance on a K8S cluster, the acme core_app is
now just generating one ACME Issuer per namespace (i.e. one
letsencrypt account per arnold project).

madmatah committed:
🔧(hello) update hello app to be compatible with k8s

We want to deploy the hello app on kubernetes. To do that, we have to
replace the Openshift specific entities with K8S equivalent :
- DeploymentConfig entities have been replaced with Deployment.
- Route entities have been replaced with Ingress

madmatah committed:
🔨(k3d) add support for RWX volumes in k3d

Some Arnold apps require PersistentVolumeClaim with ReadWriteMany
(RWX) acces mode. But out of the box, k3d is bundled with a
"local-path" StorageClass and a dynamic provisionner that can only
provision PersistentVolumes with ReadWriteOnce access mode. As a
result, our volume claims cannot be satisfied.

To solve that, since it concerns only the dev and CI environments, the
`bin/init-cluster` script pre-provision a bunch of PersistentVolume
each time we start the k3d cluster. These pre-provisionned PVs are
only virtual and do not consume or reserve any resource until they are
bound to a PVC.

The data stored on these PVs are written in a docker volume.

madmatah committed:
🔧(ashley) update ashley to be compatible with k8s

We want to use Ashley on a k8s cluster instead of Openshift.

madmatah committed:
🔧(marsha) update marsha to be compatible with k8s

We want to deploy Marsha on a k8s cluster instead of Openshift.

madmatah committed:
✨(core) allow to specify uid of running containers

By default, kubernetes will run containers with the USER specified in
the docker image. And this USER can be root. We want to run containers
as non-root user, like Openshift was doing by default.

Kubernetes allow to specify a SecurityContext for resources that will
create containers (e.g. Pods, Deployments, Jobs...). SecurityContext
has a RunAsUser option which allows to run containers with the
specified UID.

A new ansible variable `container_uid` has been introduced to specify
the UID that will be used to run containers. We recommend to specify a
unique UID per namespace.

The `hello` app has been updated to use it. And this will have to be
generalized to all arnold apps.

madmatah committed:
🚚(core) remove the concept of project in arnold

In k8s, the `Project` object type doesn't exist, it is called a
`Namespace`. All occurrences of project have been replaced with
`namespace` to avoid confusion.

madmatah committed:
🔧(redis) update redis to be compatible with k8s

We want to deploy redis on a k8s cluster instead of Openshift.

madmatah committed:
🔧(redis) speed up redis startup time

The readinessProbe declared in the redis deployment has an
initialDelaySeconds of 60s. As a result, even if the pod is started
successfully, we have to wait 60s to be able to reach it through a
Service.

It slows down deployment of other apps that depends on redis, and CI
jobs are also impacted by this.

madmatah committed:
🐛(redirect) remove labels from redirect-nginx deployment

We have a fatal error if we try to re-deploy the `redirect` core app:
the `redirect-nginx` Deployment cannot to be updated.

This is due to the fact that the `.spec.selector.matchLabels` field is
Immutable in a Deployment and that we try to patch it with different
values.

Since the `redirect` core_apps is not a blue/green application and
that the `redirect-nginx` deployment has a static name, the dynamic
labels have been removed to fix this.

madmatah committed:
🐛(redirect) add missing securityContext in redirect-nginx Deployment

The securityContext was missing in the `redirect-nginx` Deployment of
the redirect core_app.

madmatah committed:
🔧(edxapp) update edxapp to be compatible with k8s

We want to deploy edxapp on a k8s cluster instead of Openshift.

madmatah committed:
🔧(richie) update richie to be compatible with k8s

We want to deploy richie on a k8s cluster instead of Openshift.

madmatah committed:
🔧(edxec) update edxec to be compatible with k8s

We want to deploy edxec on a k8s cluster instead of Openshift.

madmatah committed:
🔧(mailcatcher) update mailcatcher to be compatible with k8s

We want to deploy mailcatcher on a k8s cluster instead of Openshift.

madmatah committed:
🔧(forum) update forum to be compatible with k8s

We want to deploy the forum app on a k8s cluster instead of Openshift.

madmatah committed:
🔧(flower) update flower to be compatible with k8s

We want to deploy flower on a k8s cluster instead of Openshift.

madmatah committed:
🔧(etherpad) update etherpad to be compatible with k8s

We want to deploy etherpad on a k8s cluster instead of Openshift.
Also, the name of the static-svc was incorrect and has been fixed.

madmatah committed:
🔧(kibana) update kibana to be compatible with k8s

We want to deploy kibana on a k8s cluster instead of Openshift.

madmatah committed:
🐛(scripts) use the right k8s context in init-cluster

The `bin/init-cluster` script executes `kubectl` commands without
specifying the configuration file related to the k3d cluster. So, the
current kubectl context is used instead. As a result, the commands
related to PV pre-provisionning and ingress-nginx controller could be
executed on another cluster.

madmatah committed:
✨(core) define a default storage class with ansible variable

A default StorageClass is not necessarily defined on a k8s cluster.
When this is the case, the PersistentVolumeClaim created by arnold are
stuck in the "Pending" state.

In order to support more use cases, new ansible variables has been
introduced to define the name of the storage class to use by default,
depending on the accessMode requested. All applications have been
updated to use them.

madmatah committed:
💬(core) replace occurences of Openshift with k8s in tasks and templates

There are still occurrences of Openshift in comments or in the names
of ansible tasks. They have been replaced by k8s.

madmatah committed:
🐛(ashley) fix postgresql deployment template

When we deploy Ashley on a non-trashable environment type, the
Deployment template of the postgresql service should be empty. But due
to a bad placement of the securityContext block, an invalid k8s entity
is generated instead.

madmatah committed:
🐛(apps) fix TLS secret name in Ingresses of blue/green compatible apps

The secret name declared in the Ingress templates of blue/green
applications is a fixed string. As a result, all Ingresses generated
with the same template are using a single secret. And that causes
conflict.

madmatah committed:
🐛(apps) fix cert-manager annotation in Ingress templates of all apps

We are using a different Issuer per Namespace to generate TLS
certificates. We have to use the `cert-manager.io/issuer` annotation
instead of `cert-manager.io/cluster-issuer`, otherwise it causes an
error.

madmatah committed:
🚨(apps) fix apiVersion of Ingresses to remove a deprecation warning

When we deploy an Ingress, we have the following warning : `Warning:
extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in
v1.22+; use networking.k8s.io/v1 Ingress`

All Ingresses' apiVersion have been updated accordingly. The new
apiVersion requires to specify a pathType for each path declared in an
Ingress. And the backend declaration structure has changed.

madmatah committed:
✨(cli) add an option to disable the use of K8S CA certificate

When using ServiceAccount discovery to authenticate against k8s API,
the CA certificate of k8s is automatically transmitted to arnold
container via the K8S_AUTH_SSL_CA_CERT_B64 environment variable.

And this CA certificate is then used by ansible to validate the
certificate of k8s' API. It is useful in situations where k8s API use
a self-signed certificate. But in other situations, the certificate
used by k8s API is signed by a public trusted CA. And we should not
verify it with k8s internal CA certificate.

A new flag has been introduced (USE_K8S_SERVICE_ACCOUNT_CA_CERT) to
disable the use of K8S CA certificate.

madmatah committed:
🐛(richie) fix postgresql deployment template

When we deploy Richie on a non-trashable environment type, the
Deployment template of the postgresql service should be empty. But due
to a bad placement of the securityContext block, an invalid k8s entity
is generated instead.

madmatah committed:
✨(cli) add a `acme` command to arnold CLI

We want to be able to update the acme Issuer easily in a namespace. A
new `acme` command has been added to arnold CLI for this purpose.

madmatah committed:
🐛(acme) fix error when the Issuer already exist in a namespace

When the create_acme.yml task is executed in a namespace that already
has its Issuer, the following error is raised : Invalid value: 0x0:
must be specified for an update.

This is due to the `force: True` attribute used during the k8s entity
creation.

madmatah committed:
🔧(apps) allow to switch the acme environment easily

We want to be able to switch the acme environment used to generate the
TLS certificate of arnold applications. It requires manual actions in
the current state of the project, due to caching issues and entities
automatically managed by cert-manager.

To make the switch process easier, we include the acme environment
(live or staging) in the name of the issuer and in the name of the TLS
secret names.

madmatah committed:
🐛(redirect) fix ingress template

The redirect ingress template did not followed updates and fixes
applied to other arnold apps templates (apiVersion, issuer annotation
and TLS secret name).

madmatah committed:
🐛(edxapp) fix type error endpoint templates

The Endpoint templates in edxapp declare target port as strings, when
it should be an integers. This causes a fatal error when we try to
deploy edxapp in a non-trashable environment type.

madmatah committed:
🎨(apps) fix code style in templates to improve readability

We want to use a consistent code style in all our templates.

madmatah committed:
🔧(redis-sentinel) update redis-sentinel to be compatible with k8s

We want to deploy the redis-sentinel app on a kubernetes cluster.
Since all the work is done by the redis-operator, the application
could have been deployed on kubernetes as-is. But a few improvements
were made : the securityContext is now consistent with the rest of the
apps and the variables redis_sentinel_user_id, redis_sentinel_group_id
and redis_sentinel_fsgroup_id have been removed. Also, the storage
class to use has been added.

madmatah committed:
🐛(cli) autodetect if we are in a terminal

We want to use arnold CLI on CI environements that don't run
necessarily inside a terminal. In these case, the docker run command
fail with the following error: "the input device is not a TTY",
because of the `-t` option.

madmatah committed:
🐛(core) trim templates before checking if we should ignore them

We are unable to deploy the forum on non-trashable environment type.
This is due to the fact that the mongodb deployment template contains
an extra-line at the end of file, and is not ignored by the "Ignore
empty rendered templates" task. It results in a fatal error during
deployment. The ansible task has been fixed to trim files and avoid
this kind of error.

madmatah committed:
🐛(elasticsearch) fix errors with secrets when security is enabled

We cannot deploy the `elasticsearch` app if the ansible variable
`elasticsearch_security_enabled` is set to false (which is the default
value). It results in a fatal error because of an undefined variable
`elasticsearch_vault_checksum`.

madmatah committed:
🔧(elasticsearch) increase memory settings for customer eugene

Elasticsearch 7.7.0 fails to start with an OutOfMemory error. The
memory heap size has been increased from 64MB to 128MB.

madmatah committed:
🔧(elasticsearch) update elasticsearch to be compatible with k8s

We want to deploy elasticsearch on a k8s cluster instead of Openshift.
The securityContext information has been added to Jobs and Statefulset
templates.

madmatah committed:
✨(core) allow to use jinja in apps defaults templates

We want be able to declare reusable variables in app defaults to avoid
code duplication.

madmatah committed:
🐛(apps) fix syntax error in nginx deployment templates

Some cleaning in commit 777d0103981714113871d364b444ecdef76387a1
introduced errors in nginx deployment template for edxapp, richie and
ashley. It can lead to a parse error when basic auth is enabled.

madmatah committed:
✨(core_apps) allow defining resources requests in redirect core_app

Resource management in pods has been added to all arnold apps, but the
redirect core app has been forgotten.

madmatah committed:
🐛(elasticsearch) fix default resources definition

When we try to deploy elasticsearch app, it fails with the following
error : `elasticsearch_app_job_set_index_template_resources` is
undefined.

madmatah committed:
✨(cli) add `pause` and `resume` commands to arnold CLI

We would like to be able to scale down an entire kubernetes namespace,
in order to save CPU and RAM resources on our development cluster.
This PR introduces 2 new commands in arnold CLI :
 - The `pause` command scales down all deployments, statefulsets and
   replication controllers that belongs to a namespace. Their previous
   replicas count is stored in an annotation.
 - The `resume` command does exactly the opposite, it restores the
   namespace in its previous state.

madmatah committed:
✨(elasticsearch) add support for persistent volume

We would like to store elasticsearch data on a persistent volume, with
a customisable storage class.

madmatah committed:
🐛(cli) fix the vaults command

Arnold CLI's `vaults` command always fails with a "No such file or
directory" error. It is due to the fact that the paths to vault files
passed to the `ansible-vault` command are absolute paths computed on
the docker host. And these paths are invalid inside the docker
container.

madmatah committed:
📝(cli) update `vaults` command description

We want to clarify the `vaults` command description.

Closes #651

lunika committed:
✨(prosody) add new prosody apps

Prosody is an XMPP server. This app is added to arnold and this is a non
blue green application. Only the http part (BOSH) of the app is exposed
outside kubernetes cluster. All services are only internal and marsha
will use there service name to access them.

jmaupetit committed:
⚡️(kibana) add kibana 7.8+ compatibility

The elasticsearch `kibana` user is now deprecated and should be named
`kibana_system` instead.

Recent kibana docker images (7.12+) no longer ship a python
distribution, breaking the status probe.

jmaupetit committed:
⚡️(scripts) add docker registry to k3d local cluster

When using arnold in a CI context for third-party apps deployable as
trays, we often have to build a Docker image of the current source code
state and test this code in a k8s context. To do so, we need a local
registry to push our newly built image that will be deployed.

To enable registry creation and attach it to the cluster, you should
define the K3D_ENABLE_REGISTRY=1 environment variable.

jmaupetit committed:
⚡️(elasticsearch) fix data volume permissions

While using RWO storage classes/providers (e.g. `cinder-csi`), PVC may
have wrong ownership and the user running in the container is not
allowed to write to the elasticsearch data volume (can belong to
`root`). Properly setting the `fsGroup` in the security context of the
pod fixes the issue.

To add more flexibility it is now also possible to also set a container
UID or GID for elasticsearch pods.

For reference, see k8s documentation:
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

sampaccoud committed:
🔧(nextcloud) update nextcloud to be compatible with k8s

We want to deploy nextcloud on a k8s cluster instead of Openshift.

sampaccoud committed:
🐛(elasticsearch) fix stateful set pvc creation deactivation

PVC can be deactivated for an Elasticsearch stateful set. In this
case the volumes should not be created.

sampaccoud committed:
🔥(forum) remove ingress now that we introduced a static service

The edxapp app was connecting to its forum via a public route because
it was the only way we initially found to make it static. We recently
introduced static services so we now use them and don't need to expose
the forum app to the internet anymore.

madmatah committed:
✨(ashley) add support for ashley 1.0.0-beta.6

Ashley 1.0.0-beta.6 introduced new settings related to media storage.
https://github.com/openfun/ashley/releases/tag/v1.0.0-beta.6

lunika committed:
🔧(marsha) update app resource requests

The resources requested for the marsha app are not accurate. Marsha
needs less cpu but more memory.

lunika committed:
✨(prosody) allow to use websocket connection

We can configure prosody to use websocket connection and/or bosh
connection. New variable are instroduce the connection to use and how to
configure them.
Also, the read_timeout connection setting can be modified.

sampaccoud committed:
✨(ingress) allow deactivating acme annotation for some prefixes

We usually don't want to expose our next/previous routes to the web
and thus can't request SLL certificates. A PR on master by @lunika
was lost when we migrated to Kubernetes. I'm porting it here.

sampaccoud committed:
🔥(flower) remove acme annotation from flower app ingress

The flower app is hidden behind an htaccess. We can't therefore request
an acme challenge and should not annotate the ingress for that.

sampaccoud committed:
🔥(docs) temporarily remove documentation awaiting revamp

The documentation is largely outdated and we don't want to delay
the release of Arnold v6 targeted to Kubernetes. We prefer to
remove all these files and bring them back in an updated version.

sampaccoud committed:
🔥(moodlenet) remove deprecated moodlenet app

The moodlenet app is not maintaied anymore and will no work
anymore in Arnold v6 after migration to kubernetes.

sampaccoud committed:
🔥(learninglocker) remove deprecated learninglocker app

The learninglocker app is not maintaied anymore and will no work
anymore in Arnold v6 after migration to kubernetes.

sampaccoud reviewed (approved):
I removed commit https://github.com/openfun/arnold/pull/620/commits/aae810f48f501a124ee20dbf1927bd2a26245bb0 from the `migration-to-k8s` and force pushed the branch.

I updated the PR description to report the work done since yesterday.

This PR should now be ready to merge.

sampaccoud committed:
🔥(tests) remove deprecated tests for moodlenet and learninglocker

When removing the moodlenet and learninglocker apps, we forgot to
remove the related tests. They make the CI fail...
